Natspec: Implement inheritance and @inheritdoc

Marenz opened the pull request:
fixes #8911

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,49 @@ using namespace solidity;
 using namespace solidity::langutil;
 using namespace solidity::frontend;
 
+namespace {
```
```suggestion
namespace
{
```

hrkrshnn commented on Changelog.md:
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 Compiler Features:
  * NatSpec: Add fields "kind" and "version" to the JSON output.
+ * NatSpec: Implement inheritance and @inheritdoc.
```
```suggestion
 * NatSpec: Implement inheritance and ``@inheritdoc.``
```

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,49 @@ using namespace solidity;
+{
+	auto& sourceDoc = dynamic_cast<StructurallyDocumentedAnnotation const&>(_source);
+
+	for (auto iterator = sourceDoc.docTags.cbegin(); iterator != sourceDoc.docTags.cend();)
+		if (_target.docTags.count(iterator->first) || iterator->first == "inheritdoc")
+			iterator++;
+		else
+			for (auto nextTag = sourceDoc.docTags.upper_bound(iterator->first); iterator != nextTag; iterator++)
+				_target.docTags.emplace(iterator->first, iterator->second);
```
I may not have understood this properly, but would the following work as a replacement?
```cpp
for (auto [key, value]: sourceDoc.docTags)
	if (!_target.docTags.count(key) && key != "inheritdoc")
		_target.docTags.emplace(key, value);
```

Marenz replied:
Unfortunately not, because it is a multimap. I there is a multi-value entry in the source, after the first insert in the target, it would no longer insert the other parts

hrkrshnn replied:
Ah, I missed that.

hrkrshnn replied:
How about we make a set of tags present in target first? Then we only check for keys from this set?

Marenz replied:
That would work, yes

Marenz replied:
I am not thrilled about the extra memory usage, but if it makes it easier to read we can do it that way

hrkrshnn replied:
I think the memory usage is minimal, since the set can contain at most 6 elements.

Marenz replied:
I have to admit, the result looks a lot nicer.

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,49 @@ using namespace solidity;
+				_target.docTags.emplace(iterator->first, iterator->second);
+}
+
+CallableDeclaration const* findBaseCallable(CallableDeclaration const& _callable, string const& _name)
+{
+	for (CallableDeclaration const* baseFuncCandidate: _callable.annotation().baseFunctions)
+		if (baseFuncCandidate->annotation().contract->name() == _name)
+			return baseFuncCandidate;
+		else if (auto callable = findBaseCallable(*baseFuncCandidate, _name))
+			return callable;
+
+	return nullptr;
+}
```
~~Doesn't this work without the "else if" part?~~

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +182,55 @@ void DocStringAnalyser::handleCallable(
+	else if (
+		_annotation.docTags.empty() &&
+		_callable.annotation().baseFunctions.size() == 1 &&
+		parameterNamesEqual(_callable, **_callable.annotation().baseFunctions.begin())
```
When comparing `_callable` and `baseFunctions`, which function should we really check? I think when we use `begin()`, the choice is almost arbitrary. I also do not know how we should make the choice.

Marenz replied:
Note that I check the size of the baseFunctions for `== 1`, which means there is only one to check and no choice to make. If there are more we don't auto-inherit ([as specified in the ticket](https://github.com/ethereum/solidity/issues/8911#issuecomment-631496750)).

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +182,55 @@ void DocStringAnalyser::handleCallable(
+		_callable.annotation().baseFunctions.size() == 1 &&
+		parameterNamesEqual(_callable, **_callable.annotation().baseFunctions.begin())
+	)
+		copyMissingTags(_annotation, (*_callable.annotation().baseFunctions.begin())->annotation());
```
The same here.

Marenz replied:
Same response here :)

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +182,55 @@ void DocStringAnalyser::handleCallable(
+		_node.documentation()->location(),
+		"Documentation tag @inheritdoc references contract \"" +
+		targetContractName +
+		"\", but the contract doesn't override this function."
```
Also when the contract doesn't exist, right?

Marenz replied:
That is true, but finding out which case it is is a bit more work which I deemed not worth the extra logic just for a slightly more detailed error message. We could rephrase it to mean both though.

hrkrshnn replied:
I think tweaking the message is fine.

hrkrshnn commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -1214,6 +1214,482 @@ BOOST_AUTO_TEST_CASE(slash3_slash4)
+
+}
+
+BOOST_AUTO_TEST_SUITE_END()
```
Can you add a test where a state variable overrides a function? Something like this:

```solidity
  contract C {
	  /// @notice Hello world
	  function x() virtual external returns (uint) {
		  return 1;
	  }
  }

  contract D is C {
	  uint public override x;
  }
```

hrkrshnn commented:
We are missing an entry in the documentation for `@inheritdoc`. Other than that, this looks good.

chriseth commented:
Does this both implement the explicit inheritance via `@inheritdoc` and the implicit inheritance or just the first?

Marenz commented:
@chriseth this implements both

hrkrshnn commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,52 @@ using namespace solidity;
+
+CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFuncs, string const& _name)
+{
+	for (CallableDeclaration const* baseFuncCandidate: _baseFuncs)
+		if (baseFuncCandidate->annotation().contract->name() == _name)
+			return baseFuncCandidate;
+		else if (auto callable = findBaseCallable(baseFuncCandidate->annotation().baseFunctions, _name))
+			return callable;
+
+	return nullptr;
```
I wonder if it's possible to pick the contract first from  `ContractAnnotation::linearizedBaseContracts`, followed by the function? What do you think?

Marenz replied:
What advantages do you hope to gain from that? If we use the contracts as basis (instead of the already deduced baseFunctions in the annotation), we still have to match the function name & parameters of each contract with this one, thus redoing work that was already done to get the content of `baseFunctions`..

hrkrshnn replied:
A minor advantage is that we can distinguish between "contract doesn't exist" and "contract doesn't contain a function that we want." But I agree that matching the function after finding the contract might be tricky.

hrkrshnn commented:
There is a minor issue. According to [documentation](https://solidity.readthedocs.io/en/v0.6.10/natspec-format.html), `@author` is not applicable to public state variables, but applicable to functions. What should happen if a public state variable overrides a function that has an `@author` tag?

This could be an issue with the documentation, because when running tests, I was able to use `@author` on public state variables.

christianparpart commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,52 @@ using namespace solidity;
+
+	set<string> existingTags;
+
+	for (auto& iterator: _target.docTags)
```
```suggestion
	for (auto const& iterator: _target.docTags)
```

christianparpart commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,52 @@ using namespace solidity;
+	for (auto& iterator: _target.docTags)
+		existingTags.insert(iterator.first);
+
+	for (auto& [tag, content]: sourceDoc.docTags)
```
```suggestion
	for (auto const& [tag, content]: sourceDoc.docTags)
```

christianparpart commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,52 @@ using namespace solidity;
+	std::vector<ASTPointer<VariableDeclaration>> const& aParams = _a.parameters();
+	std::vector<ASTPointer<VariableDeclaration>> const& bParams = _b.parameters();
+
+	return equal(
```
Try using boost::ranges::equal(...) here, for example like:
```suggestion
	return boost::range::equal(_a.parameters(), _b.parameters(), [](auto const& pa, auto const& pb) { return pa->name() == pb->name(); };
```

To effectively making this implementation a one-liner.

Marenz replied:
I thought we try to avoid boost :P

christianparpart commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +194,55 @@ void DocStringAnalyser::handleCallable(
+
+	string const targetContractName = _annotation.docTags.find("inheritdoc")->second.content;
+
+	if (auto callable = findBaseCallable(_baseFuncs, targetContractName))
```
```suggestion
	if (auto const callable = findBaseCallable(_baseFuncs, targetContractName))
```

christianparpart reviewed (changes requested):
Mostly constness suggestions and one boost algorithm proposal.

Marenz commented:
@hrkrshnn according to the code `author` is fine for public vars. I'll fix the docu

Marenz commented:
Ah well.. no.

```
		if (_variable.annotation().docTags.count("title") > 0 || _variable.annotation().docTags.count("author") > 0)
			m_errorReporter.warning(
				8532_error, _variable.documentation()->location(),
				"Documentation tag @title and @author is only allowed on contract definitions. "
				"It will be disallowed in 0.7.0."
			);
```

Marenz commented:
Though we don't have the same warning for functions..

Marenz commented:
So, I couldn't find a concrete source, but the warning about using `author` was introduced by https://github.com/ethereum/solidity/issues/3418 and the comments there read like the intention is that it will be disallowed on functions as well and the warning was just forgotten for functions. 

I also found an incorrect piece in the natspec docu that I removed in a second commit here. I also added that warning for functions now.

@chriseth @aarlt  please confirm.

christianparpart commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,62 @@ void DocStringAnalyser::handleCallable(
+{
+	size_t const occurrences = _annotation.docTags.count("inheritdoc");
+
+	if (occurrences == 0)
```
Here you have exactly three cases. either 0, 1 or else. I'd suggest to use a switch statement for that to better express the intention.

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `a68f5762ad44a612bf9b548735b512700bcaf82e`:
```
Coding style error:
 libsolidity/analysis/DocStringAnalyser.cpp:213: switch(_annotation.docTags.count("inheritdoc"))

```
Please check that your changes are working as intended.

stackenbotten commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,59 @@ void DocStringAnalyser::handleCallable(
+	StructurallyDocumentedAnnotation& _annotation
+)
+{
+	switch(_annotation.docTags.count("inheritdoc"))
```
Coding style error

christianparpart reviewed (dismissed):
LGTM. :)

chriseth commented on docs/natspec-format.rst:
```diff
@@ -65,11 +65,18 @@ The following example shows a contract and a function using all available tags.
     }
 
+    contract SubTree is Tree {
+        /// @inheritdoc Tree
```
Isn't this a bad example because we would not need the `@inheritdoc` here? Can you add another function where it is needed and then say that the first inherits, the second needs an explicit tag?

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,41 @@ using namespace solidity;
+
+	set<string> existingTags;
+
+	for (auto const& iterator: _target.docTags)
```
`_target.docTags` is a `multimap` - do we really need another set for the existing keys?

Marenz replied:
Exactly because it is a multimap we do need this. If we add a multi-value that didn't previously exist to `_target`, we would not add the remaining values of that entry if we just check for existence of the key.

chriseth replied:
ah, right! Thanks for the explanation!

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -32,6 +32,41 @@ using namespace solidity;
+CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFuncs, string const& _name)
+{
+	for (CallableDeclaration const* baseFuncCandidate: _baseFuncs)
+		if (baseFuncCandidate->annotation().contract->name() == _name)
```
If we do it properly, this should do actual reference resolution - would that be very complicated?
The reason I'm bringing this up is that this name might not be the name under which the contract is known in the current file.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,52 @@ void DocStringAnalyser::handleCallable(
+{
+	switch (_annotation.docTags.count("inheritdoc"))
+	{
+		default:
```
We do not indent switch case names

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,52 @@ void DocStringAnalyser::handleCallable(
+		case 0:
+			return nullptr;
+		case 1:
+			string const targetContractName = _annotation.docTags.find("inheritdoc")->second.content;
```
This can contain spaces, can it?

Marenz replied:
I did a short test and it seems it's trimmed. A space before/after is ignored.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,52 @@ void DocStringAnalyser::handleCallable(
+				_node.documentation()->location(),
+				"Documentation tag @inheritdoc references contract \"" +
+				targetContractName +
+				"\", but the contract doesn't exist or doesn't override this function."
```
```suggestion
				"\", but the contract does not exist or is not a base contract that defines this function."
```

chriseth commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -1214,6 +1214,638 @@ BOOST_AUTO_TEST_CASE(slash3_slash4)
+	char const *natspec = R"ABCDEF({
+		"methods":
+		{
+			"x()":
```
I think it would be easier to grasp the meaning of the test if this was just a single line

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -139,9 +183,52 @@ void DocStringAnalyser::handleCallable(
 	checkParameters(_callable, _node, _annotation);
+
+	if (CallableDeclaration const* baseFunction = resolveInheritDoc(_callable.annotation().baseFunctions, _node, _annotation))
+		copyMissingTags(_annotation, baseFunction->annotation());
```
Does `copyMissingTags` depend on the order in which the contracts are defined in the source file?
What about: A inherits from B inherits from C
```
contract A is B{}
contract B is C {}
contract C {}
```
And `B` explicitly inherits the docs from `C`, but `A` implicitly inherits the docs from `B`?

Marenz replied:
Yes it depends on the order. But your example doesn't work:
```
Error: Definition of base has to precede definition of derived contract
 --> test.sol:1:15:
  |
1 | contract A is B{}
  |               ^


```

chriseth replied:
Ah, so you are lucky! :)

chriseth replied:
Can you still add a test where the middle contract inherits explicitly and the most derived inherits implicitly?

Marenz committed:
NatSpec: Implement ``@inheritdoc``

stackenbotten commented:
There was an error when running `chk_coding_style` for commit `0b88d50e73d43ee2a2f22bfd658df198f84a5ca8`:
```
Coding style error:
 libsolidity/analysis/ReferencesResolver.cpp:304: switch(_annotation.docTags.count("inheritdoc"))

```
Please check that your changes are working as intended.

stackenbotten commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +299,57 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
 
+void ReferencesResolver::resolveInheritDoc(StructuredDocumentation const& _documentation, StructurallyDocumentedAnnotation& _annotation)
+{
+	switch(_annotation.docTags.count("inheritdoc"))
```
Coding style error

chriseth commented on Changelog.md:
```diff
@@ -7,6 +7,7 @@ Language Features:
 Compiler Features:
  * NatSpec: Add fields "kind" and "version" to the JSON output.
  * NatSpec: Inherit tags from unique base if derived function does not provide any.
+ * NatSpec: Implement ``@inheritdoc``.
```
Please explain in a few words what it does.

chriseth commented on docs/natspec-format.rst:
```diff
@@ -49,7 +49,7 @@ The following example shows a contract and a function using all available tags.
 .. code:: solidity
```
I think the statements in the previous paragraph are wrong.

Marenz replied:
A short test confirmed that it is correct. Probably because we use `interfaceFunctions()` to iterate over the things that will eventually be output as natspec (see Natspec.cpp)

chriseth replied:
We do parse tags on internal functions.

Marenz replied:
ah. I guess we check them but we never output anything for it

chriseth commented on docs/natspec-format.rst:
```diff
@@ -60,9 +60,34 @@ The following example shows a contract and a function using all available tags.
+        }
+    }
+
+    contract Plant {
```
I think a Tree should be a Plant...

chriseth commented on docs/natspec-format.rst:
```diff
@@ -60,9 +60,34 @@ The following example shows a contract and a function using all available tags.
+        }
+    }
+
+    contract SubTree is Tree, Plant {
```
Maybe use the name of a specific kind of tree?

chriseth commented on docs/natspec-format.rst:
```diff
@@ -60,9 +60,34 @@ The following example shows a contract and a function using all available tags.
+            return rings + 2;
+        }
+
+        /// For multiple inheritance we need to explicitly specify the base
```
This sentence goes into the `@notice` tag, so it should describe what the function does.

By the way, do we have a test that checks that leaving out the `@notice` still works the same way with default and explicit inheritance?

Marenz replied:
Yes we have such a test, see `dev_explicit_inherit`

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -52,6 +52,17 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
 			_target.docTags.emplace(tag, content);
 }
 
+CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFuncs, ContractDefinition const& _contract)
```
```suggestion
CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFunctions, ContractDefinition const& _contract)
```

chriseth replied:
I think we should not use abbreviation for function parameters.

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -52,6 +52,17 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
+CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFuncs, ContractDefinition const& _contract)
+{
+	for (CallableDeclaration const* baseFuncCandidate: _baseFuncs)
+		if (baseFuncCandidate->annotation().contract->id() == _contract.id())
```
Would it be enough to specify the contract id as parameter?

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +299,58 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+		m_errorReporter.docstringParsingError(
+			5142_error,
+			_documentation.location(),
+			"Documentation tag @inheritdoc can only reference one contract."
```
Strictly, all tags could refer to the same contract. So maybe: ... can only be given once?

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +299,58 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
 
+void ReferencesResolver::resolveInheritDoc(StructuredDocumentation const& _documentation, StructurallyDocumentedAnnotation& _annotation)
+{
+	switch (_annotation.docTags.count("inheritdoc"))
```
Do you think using a switch makes this easy to understand? :)

Marenz replied:
At first I wasn't but @christianparpart  convinced me and now I think he's right.

chriseth replied:
So you both think that upon reading the world `default` it is fine to be forced to first check what the other cases are to arrive at the conclusion that the default case is "more than one"?

Marenz replied:
To be fair, the position of the `default:` I'd rather have had at the end of the cases, but the compiler didn't like that

Marenz replied:
Ah, now it does. Dunno what the problem was

chriseth commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -78,6 +78,8 @@ struct StructurallyDocumentedAnnotation
 	/// Mapping docstring tag name -> content.
 	std::multimap<std::string, DocTag> docTags;
+	/// contract that @inheritdoc references if it exists
+	ContractDefinition const* inheritContract = nullptr;
```
Maybe
```suggestion
	ContractDefinition const* inheritdocReference = nullptr;
```
?

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -362,6 +368,8 @@ bool CompilerStack::analyze()
 						if (!contractLevelChecker.check(*contract))
 							noErrors = false;
 
+		// Requires annotation().baseFuncs which is filled by
```
I think this comment is way too specific to be put into the compiler stack.

chriseth commented on libsolidity/analysis/DocStringTagParser.h:
```diff
@@ -0,0 +1,81 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * @author Christian <c@ethdev.com>
+ * @date 2015
+ * Parses and analyses the doc strings.
+ * Stores the parsing results in the AST annotations and reports errors.
```
```suggestion
 * Parses the doc strings.
```

chriseth replied:
Or only their tags?

chriseth commented on libsolidity/analysis/DocStringTagParser.h:
```diff
@@ -0,0 +1,81 @@
+class DocStringTagParser: private ASTConstVisitor
+{
+public:
+	DocStringTagParser(langutil::ErrorReporter& _errorReporter): m_errorReporter(_errorReporter) {}
```
explicit

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -52,6 +52,17 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
 			_target.docTags.emplace(tag, content);
 }
 
+CallableDeclaration const* findBaseCallable(std::set<CallableDeclaration const*>& _baseFunctions, int64_t _contractId)
```
```suggestion
CallableDeclaration const* findBaseCallable(set<CallableDeclaration const*> const& _baseFunctions, int64_t _contractId)
```

chriseth commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -127,133 +111,41 @@ bool DocStringAnalyser::visit(EventDefinition const& _event)
+		_node.documentation()->location(),
+		"Documentation tag @inheritdoc references contract \"" +
+		_annotation.inheritdocReference->name() +
+		"\", but the contract does not override this function."
```
shouldn't it be
"...contract does not contain a function that is overridden by this function"?

chriseth commented on libsolidity/analysis/DocStringTagParser.h:
```diff
@@ -0,0 +1,75 @@
+{
+
+/**
+ * Parses the doc tags and does basic vadility checks.
```
```suggestion
 * Parses the doc tags and does basic validity checks.
```

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +299,61 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+				5142_error,
+				_documentation.location(),
+				"Documentation tag @inheritdoc can only be given once."
+				);
```
indentation

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +299,61 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+	case 1:
+	{
+		string const& name = _annotation.docTags.find("inheritdoc")->second.content;
+		vector<Declaration const*> results = m_resolver.nameFromCurrentScope(name);
```
I think we should use `pathFromCurrentScope` to allow more complex use-cases.

chriseth commented:
Good apart from the comments!

chriseth commented on Changelog.md:
```diff
@@ -26,6 +26,7 @@ Language Features:
 Compiler Features:
  * NatSpec: Add fields ``kind`` and ``version`` to the JSON output.
  * NatSpec: Inherit tags from unique base functions if derived function does not provide any.
+ * NatSpec: Implement tag ``@inheritdoc`` to copy documentation from a specific contract.
```
Wrong version

chriseth replied:
not yet fixed.

Marenz replied:
I only rebased, fixes are about to come

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +300,54 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+		string const& name = _annotation.docTags.find("inheritdoc")->second.content;
+		vector<string> path;
+		boost::split(path, name, boost::is_any_of("."));
+		Declaration const* result = m_resolver.pathFromCurrentScope(path);
```
Do you have a test case for that? You will need to use a multi-file test that uses import.

Marenz replied:
This test does a resolving of a path: https://github.com/ethereum/solidity/pull/9218/files#diff-5dcd49ff3e08e5ff224ee08fbb7ac514

It's not a contract and thus fails later but it does successfully resolve the path.

chriseth replied:
Can you please still add a multi-file test?

leonardoalt commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -52,6 +52,17 @@ void copyMissingTags(StructurallyDocumentedAnnotation& _target, set<CallableDecl
 			_target.docTags.emplace(tag, content);
 }
 
+CallableDeclaration const* findBaseCallable(set<CallableDeclaration const*> const& _baseFunctions, int64_t _contractId)
```
Why `int`?

Marenz replied:
Because that's the type that `ContractDefinition->id()` returns

leonardoalt replied:
Indeed. Now I'm confused why `id` is int

leonardoalt commented on libsolidity/analysis/DocStringAnalyser.cpp:
```diff
@@ -127,133 +111,41 @@ bool DocStringAnalyser::visit(EventDefinition const& _event)
 
-void DocStringAnalyser::parseDocStrings(
+CallableDeclaration const* DocStringAnalyser::resolveInheritDoc(
+	std::set<CallableDeclaration const*>& _baseFuncs,
```
```suggestion
	std::set<CallableDeclaration const*> const& _baseFuncs,
```
?

leonardoalt commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +300,54 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+			m_errorReporter.docstringParsingError(
+				9397_error,
+				_documentation.location(),
+				"Documentation tag @inheritdoc references unexisting contract \"" +
```
```suggestion
				"Documentation tag @inheritdoc references inexistent contract \"" +
```

leonardoalt commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -1318,15 +1312,82 @@ BOOST_AUTO_TEST_CASE(user_default_inherit_variable)
+	checkNatspec(sourceCode, "D", natspec, true);
+}
+
+BOOST_AUTO_TEST_CASE(dev_explicit_inherit_variable)
```
Did these tests have to be added via cpp tests? Couldn't they have been added as command line tests? (or something that is not cpp tests)

Marenz replied:
I suppose anything could be a cmdline test, but this file is specifically for natspecs so it does seem to be the right place..
Whether we want/should turn the whole thing into not-cpp tests is a different question (of which I'd be in favour)

leonardoalt replied:
Yea, I just wondered if some similar tests had already been "converted" into cmd line tests

Marenz replied:
I am not sure cmd line tests should be the favored target, maybe a more specific extension in (i)soltest would be better.

leonardoalt commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -57,28 +57,22 @@ Json::Value Natspec::userDocumentation(ContractDefinition const& _contractDef)
-					Json::Value user;
-					// since @notice is the only user tag if missing function should not appear
-					user["notice"] = Json::Value(value);
-					methods[it.second->externalSignature()] = user;
```
What happened to these 2?

Marenz replied:
I simplified the code a bit, they just moved to line 74/75

leonardoalt commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -283,4 +300,54 @@ void ReferencesResolver::operator()(yul::VariableDeclaration const& _varDecl)
+				);
+	}
+	break;
+}
```
Indentation looks off

leonardoalt commented on libsolidity/analysis/DocStringTagParser.cpp:
```diff
@@ -0,0 +1,228 @@
+				"Documentation tag @" + docTag.first + " not valid for " + _nodeName + "."
+			);
+		else
+			if (docTag.first == "return")
```
```suggestion
		else if (docTag.first == "return")
```

leonardoalt commented on libsolidity/analysis/DocStringTagParser.cpp:
```diff
@@ -0,0 +1,228 @@
+			if (docTag.first == "return")
+			{
+				returnTagsVisited++;
+				if (auto* varDecl = dynamic_cast<VariableDeclaration const*>(&_node))
```
```suggestion
				if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(&_node))
```

leonardoalt commented on libsolidity/analysis/DocStringTagParser.cpp:
```diff
@@ -0,0 +1,228 @@
+							"Documentation tag \"@" + docTag.first + "\" is only allowed once on state-variables."
+						);
+				}
+				else if (auto* function = dynamic_cast<FunctionDefinition const*>(&_node))
```
```suggestion
				else if (auto const* function = dynamic_cast<FunctionDefinition const*>(&_node))
```
